### PR TITLE
[primTorch] Make `prims.collapse` a real prim

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92017
* __->__ #91748

`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

cc @ezyang @mruberry @ngimel @Lezcano

### CODE REVIEW

**Comment by ezyang** [comment]
mmm, i can review this in a pinch but I think the other ccs will be better for it

File: torch/_prims/__init__.py [hunk 1]
```diff
@@ -1832,19 +1833,38 @@ def _as_strided_scatter_meta(
...
-    """
 
-    dim_length = 1
-    for idx in range(start, end):
-        dim_length = dim_length * a.shape[idx]
+def _collapse_meta(a: Tensor, start: int, end: int) -> Tensor:
+    # Special-case for zero dimensional tensors
+    ndim = max(1, a.dim())
+    utils.validate_idx(ndim, start)
+    utils.validate_exclusive_idx(ndim, end)
```
Comment (mruberry) [inline torch/_prims/__init__.py:1842]: Won't this prevent `collapse(t, 0, 0)` - is that invalid?

File: torch/_prims/__init__.py [hunk 2]
```diff
@@ -1832,19 +1833,38 @@ def _as_strided_scatter_meta(
...
 
-    new_shape = a.shape[0:start] + (dim_length,) + a.shape[end:]
-    return reshape(a, new_shape)
+    new_shape = _collapsed_shape(a.shape, start, end)
```
Comment (mruberry) [inline torch/_prims/__init__.py:1844]: Is there a check that `end >= start`?

File: torch/_prims/__init__.py [hunk 3]
```diff
@@ -1832,19 +1833,38 @@ def _as_strided_scatter_meta(
...
+        out.view_as(a).copy_(a)
+    return out
+
+
+_collapse_doc = """
+Collapse a span of neighboring dimensions into one.
+
+See collapse_view for the corresponding view operation.
+"""
+collapse = _make_prim(
```
Comment (mruberry) [inline torch/_prims/__init__.py:1861]: Is there any testing for collapse? Should we add some tests to test_prims.py?

File: torch/_prims/__init__.py (hunk 1 above)
Comment (peterbell10) [reply torch/_prims/__init__.py:1842]: For whatever reason `aten.flatten(t, 0, 1)` flattens dims [0, 1] inclusive while `prims.collapse(t, 0, 1)` collapses dims `[0, 1)` exclusive. So the range [0, 0) is empty which is not allowed.

Comment (ngimel) [reply torch/_prims/__init__.py:1842]: Why does behavior of `prims.collapse` and `aten.flatten` differ? We are under no obligation to follow `lax` names and behavior, or are we?

Comment (peterbell10) [reply torch/_prims/__init__.py]: Looks like `prims.collapse` has been this way since the first prims PR #75095. cc @mruberry

Comment (mruberry) [reply torch/_prims/__init__.py]: We didn't intend to fix the prims' from day one, I was just modeling them on lax at the time to get the project started

File: torch/_prims/__init__.py (hunk 2 above)
Comment (peterbell10) [reply torch/_prims/__init__.py:1854]: Added `_validate_collapse_args` which now handles this check.

File: torch/_prims/__init__.py (hunk 3 above)
Comment (peterbell10) [reply torch/_prims/__init__.py:1871]: Test added

**Comment by peterbell10** [comment]
@mruberry PTAL

**Comment by peterbell10** [comment]
@mruberry @ngimel @lezcano ping

File: torch/_prims/__init__.py [hunk 4]
```diff
@@ -1335,25 +1355,12 @@ def _collapse_view_meta(a: TensorLikeType, start: int, end: int) -> TensorLikeTy
         msg = "Attempting to view a collapsed tensor, but no such view exists!"
         raise ValueError(msg)
 
-    if new_strides is None:
-        return a.view(new_shape)
-    else:
-        return a.as_strided(new_shape, new_strides, a.storage_offset())
+    assert new_strides is not None
```
Comment (lezcano) [inline torch/_prims/__init__.py:1358]: Let's just raise within `collapse_view_helper` and never return `None`.

**Review by lezcano — APPROVED** [review]

File: torch/_prims/__init__.py (hunk 4 above)
Comment (peterbell10) [reply torch/_prims/__init__.py:1358]: `_collapse_view_helper` is used by `_refs.flatten` which returns a copy when `_collapse_view_helper` returns `None`.

Comment (lezcano) [reply torch/_prims/__init__.py:1358]: oof. ok

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**